From 54381dafa4613dea6bb1e1b9d9b9f744181876fb Mon Sep 17 00:00:00 2001
From: Bruno Martins <bgcngm@gmail.com>
Date: Sun, 30 Dec 2018 16:02:59 +0000
Subject: [PATCH 19/47] AppOpsService: Watch op mode changes when an AppOp
 restriction dies

On binder death, ensure that any associated watched ops are not left
dangling in an active state.

Change-Id: I060d9fabc555e609ad529c612fcb7712f0cea0a6
---
 services/core/java/com/android/server/AppOpsService.java | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/AppOpsService.java b/services/core/java/com/android/server/AppOpsService.java
index 20d0e20a270..cdee2ba49c1 100644
--- a/services/core/java/com/android/server/AppOpsService.java
+++ b/services/core/java/com/android/server/AppOpsService.java
@@ -611,7 +611,11 @@ public class AppOpsService extends IAppOpsService.Stub {
         public void binderDied() {
             synchronized (AppOpsService.this) {
                 for (int i=mStartedOps.size()-1; i>=0; i--) {
-                    finishOperationLocked(mStartedOps.get(i), /*finishNested*/ true);
+                    final Op op = mStartedOps.get(i);
+                    finishOperationLocked(op, /*finishNested*/ true);
+                    if (op.startNesting <= 0) {
+                        scheduleOpActiveChangedIfNeededLocked(op.op, op.uid, op.packageName, false);
+                    }
                 }
                 mClients.remove(mAppToken);
             }
-- 
2.17.1

